### PR TITLE
Sum Grad Sqaured and Variance extensions

### PR DESCRIPTION
**Pull request by sbharadwajj**
1. Extensions for `SumGradSquared` and `Variance` for the layers: `Conv1d`, `Conv3d`, `ConvTranspose1,2,3d`
2. Test suite for the same

Note: Change of constant name from `N_axis` to `C_axis` for the `Conv` layers for better readability. Can be found in all the `conv*.py`

### CODE REVIEW

File: backpack/extensions/firstorder/sum_grad_squared/__init__.py [hunk 1]
```diff
@@ -1,20 +1,33 @@
...
 class SumGradSquared(BackpropExtension):
     """The sum of individual-gradients-squared, or second moment of the gradient.
-
     Stores the output in ``sum_grad_squared``. Same dimension as the gradient.
-
     Note: beware of scaling issue
         The second moment depends on the scaling of the overall function.
         Let ``fᵢ`` be the loss of the ``i`` th sample, with gradient ``gᵢ``.
         ``SumGradSquared`` will return the sum of the squared
-
```
Comment (f-dangel) [inline backpack/extensions/firstorder/sum_grad_squared/__init__.py:17]: Was there a reason why these lines had to go? I think they are necessary to prettify the [docs](https://docs.backpack.pt/en/master/extensions.html#backpack.extensions.SumGradSquared)

File: backpack/extensions/firstorder/sum_grad_squared/conv1d.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+from backpack.extensions.firstorder.base import FirstOrderModuleExtension
+from backpack.utils import conv as convUtils
+
+
+class SGSConv1d(FirstOrderModuleExtension):
+    def __init__(self):
+        super().__init__(params=["bias", "weight"])
+
+    def bias(self, ext, module, g_inp, g_out, backproped):
+        C_axis = 0
```
Comment (f-dangel) [inline backpack/extensions/firstorder/sum_grad_squared/conv1d.py:12]: The 0.th dimension of `g_out` is the batch dimension. Rename to `N_axis`.

File: backpack/extensions/firstorder/sum_grad_squared/conv1d.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        return (einsum("ncl->nc", g_out[0]) ** 2).sum(C_axis)
+
+    def weight(self, ext, module, g_inp, g_out, backproped):
+        C_axis = 0
```
Comment (f-dangel) [inline backpack/extensions/firstorder/sum_grad_squared/conv1d.py:16]: 0th dimension should be `N_axis`

File: backpack/extensions/firstorder/sum_grad_squared/conv2d.py [hunk 4]
```diff
@@ -9,13 +9,13 @@ def __init__(self):
         super().__init__(params=["bias", "weight"])
 
     def bias(self, ext, module, g_inp, g_out, backproped):
-        N_axis = 0
-        return (einsum("nchw->nc", g_out[0]) ** 2).sum(N_axis)
+        C_axis = 0
+        return (einsum("nchw->nc", g_out[0]) ** 2).sum(C_axis)
```
Comment (f-dangel) [inline backpack/extensions/firstorder/sum_grad_squared/conv2d.py:13]: Please revert

File: backpack/extensions/firstorder/sum_grad_squared/conv2d.py [hunk 5]
```diff
@@ -9,13 +9,13 @@ def __init__(self):
...
 
     def weight(self, ext, module, g_inp, g_out, backproped):
-        N_axis = 0
+        C_axis = 0
         X, dE_dY = convUtils.get_weight_gradient_factors(
             module.input0, g_out[0], module
         )
         d1 = einsum("nml,nkl->nmk", (dE_dY, X))
-        return (d1 ** 2).sum(N_axis).view_as(module.weight)
+        return (d1 ** 2).sum(C_axis).view_as(module.weight)
```
Comment (f-dangel) [inline backpack/extensions/firstorder/sum_grad_squared/conv2d.py:21]: Please revert and assign a name to  `d1` that is easier to understand.

File: backpack/extensions/firstorder/sum_grad_squared/conv3d.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+        C_axis = 0
+        return (einsum("ncdhw->nc", g_out[0]) ** 2).sum(C_axis)
+
+    def weight(self, ext, module, g_inp, g_out, backproped):
+        C_axis = 0
+        X, dE_dY = convUtils.get_conv3d_weight_gradient_factors(
+            module.input0, g_out[0], module
+        )
+        d1 = einsum("nml,nkl->nmk", (dE_dY, X))
+        return (d1 ** 2).sum(C_axis).view_as(module.weight)
```
Comment (f-dangel) [inline backpack/extensions/firstorder/sum_grad_squared/conv3d.py:21]: same as above

File: backpack/extensions/firstorder/sum_grad_squared/convtranspose1d.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        C_axis = 0
+        C_out_axis = 1
+
+        X, dE_dY = convUtils.get_convtranspose1d_weight_gradient_factors(
+            module.input0, g_out[0], module
+        )
+        C_in, C_out, K_X = module.weight.shape
+        d1 = einsum("nml,nkl->nmk", (dE_dY, X))
+        d2 = (d1 ** 2).sum(C_axis).reshape(C_out, C_in, K_X)
+        return transpose(d2, C_axis, C_out_axis)
```
Comment (f-dangel) [inline backpack/extensions/firstorder/sum_grad_squared/convtranspose1d.py:25]: same as above. Re axes: We sum out the batch dimension (`N_axis`). Before that, the axis of out channels is `1`. After summing over the batch dimension, the axis of out channels is `0` (because the batch dimension has been summed out). This is currently not correctly represented by the names.

File: backpack/extensions/firstorder/sum_grad_squared/convtranspose2d.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+    def weight(self, ext, module, g_inp, g_out, backproped):
+        C_axis = 0
+        C_out_axis = 1
+        X, dE_dY = convUtils.get_convtranspose2d_weight_gradient_factors(
+            module.input0, g_out[0], module
+        )
+        C_in, C_out, K_X, K_Y = module.weight.shape
+        d1 = einsum("nml,nkl->nmk", (dE_dY, X))
+        d2 = (d1 ** 2).sum(C_axis).reshape(C_out, C_in, K_X, K_Y)
+        return transpose(d2, C_axis, C_out_axis)
```
Comment (f-dangel) [inline backpack/extensions/firstorder/sum_grad_squared/convtranspose2d.py:24]: same as above

File: backpack/extensions/firstorder/sum_grad_squared/convtranspose3d.py [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    def weight(self, ext, module, g_inp, g_out, backproped):
+        C_axis = 0
+        C_out_axis = 1
+        X, dE_dY = convUtils.get_convtranspose3d_weight_gradient_factors(
+            module.input0, g_out[0], module
+        )
+        C_in, C_out, K_X, K_Y, K_Z = module.weight.shape
+        d1 = einsum("nml,nkl->nmk", (dE_dY, X))
+        d2 = (d1 ** 2).sum(C_axis).reshape(C_out, C_in, K_X, K_Y, K_Z)
+        return transpose(d2, C_axis, C_out_axis)
```
Comment (f-dangel) [inline backpack/extensions/firstorder/sum_grad_squared/convtranspose3d.py:24]: same as above

File: test/extensions/firstorder/batch_grad/batchgrad_settings.py [hunk 10]
```diff
@@ -1,142 +1,10 @@
...
+The tests are taken from `test.extensions.firstorder.firstorder_settings`, 
+but additional custom tests can be defined here by appending it to the list.
 """
 
-
-import torch
-from test.core.derivatives.utils import classification_targets, regression_targets
+from test.extensions.firstorder.firstorder_settings import FIRSTORDER_SETTINGS
 
 BATCHGRAD_SETTINGS = []
```
Comment (f-dangel) [inline test/extensions/firstorder/batch_grad/batchgrad_settings.py:9]: A remark about naming: I would prefer if we had something like
```
SHARED_SETTINGS = FIRSTORDER_SETTINGS
LOCAL_SETTINGS = [...] # could be empty
BATCHGRAD_SETTINGS = SHARED_SETTINGS + LOCAL_SETTINGS
```
Can you impose a similar convention for all firstorder extensions in the test suite?

File: test/extensions/firstorder/firstorder_settings.py [hunk 11]
```diff
@@ -0,0 +1,153 @@
+"""Test configurations for `backpack.core.extensions.firstorder` methods
+
+
```
Comment (f-dangel) [inline test/extensions/firstorder/firstorder_settings.py:3]: Can you add a comment that these settings will be shared among all first-order executions?

File: test/extensions/firstorder/variance/test_variance.py [hunk 12]
```diff
@@ -0,0 +1,35 @@
+"""Test class for module variance 
+from `backpack.core.extensions.firstorder`
+
+Test individual gradients for the following layers:
```
Comment (f-dangel) [inline test/extensions/firstorder/variance/test_variance.py:4]: should be variances

File: backpack/extensions/firstorder/sum_grad_squared/__init__.py (hunk 1 above)
Comment (sbharadwajj) [reply backpack/extensions/firstorder/sum_grad_squared/__init__.py:17]: I think they were removed during `black`. I dont recall removing lines

File: backpack/extensions/firstorder/sum_grad_squared/convtranspose1d.py (hunk 7 above)
Comment (sbharadwajj) [reply backpack/extensions/firstorder/sum_grad_squared/convtranspose1d.py:25]: Ah right. You are correct that we sum out the batch dimension everywhere (`N_axis`).

So it should be `transpose(d2, C_out_axis, C_in_axis)`, right?
Intuitively atleast, not code wise.

File: backpack/extensions/firstorder/sum_grad_squared/__init__.py (hunk 1 above)
Comment (sbharadwajj) [reply backpack/extensions/firstorder/sum_grad_squared/__init__.py:17]: fixed
